qp_klp/StandardMetagenomicWorkflow.py

```diff
@@ -0,0 +1,184 @@
+        super().__init__(**kwargs)
+
+        self.mandatory_attributes = ['qclient', 'uif_path',
+                                     'lane_number', 'config_fp',
+                                     'run_identifier', 'output_dir', 'job_id',
+                                     'lane_number', 'is_restart']
+
+        self.confirm_mandatory_attributes()
+
+        # second stage initializer that could conceivably be pushed down into
+        # specific children requiring specific parameters.
+        self.qclient = self.kwargs['qclient']
+
+        self.pipeline = Pipeline(self.kwargs['config_fp'],
+                                 self.kwargs['run_identifier'],
+                                 self.kwargs['uif_path'],
+                                 self.kwargs['output_dir'],
+                                 self.kwargs['job_id'],
+                                 ASSAY_NAME_METAGENOMIC,
+                                 lane_number=self.kwargs['lane_number'])
+
+        self.fsr = FailedSamplesRecord(self.kwargs['output_dir'],
+                                       self.pipeline.sample_sheet.samples)
+
+        self.master_qiita_job_id = None
+
+        self.lane_number = self.kwargs['lane_number']
+        self.is_restart = bool(self.kwargs['is_restart'])
+
+        if self.is_restart is True:
+            self.determine_steps_to_skip()
+
+        # this is a convenience member to allow testing w/out updating Qiita.
+        self.update = True
+
+        if 'update_qiita' in kwargs:
+            if bool(kwargs['update_qiita']) is False:
+                self.update = False
+
+        # for now, hardcode this at the legacy value, since we've never
+        # changed it.
+        self.job_pool_size = 30
+
+    def determine_steps_to_skip(self):
+        out_dir = self.pipeline.output_path
+
+        directories_to_check = ['ConvertJob', 'NuQCJob',
+                                'FastQCJob', 'GenPrepFileJob']
+
+        for directory in directories_to_check:
+            if exists(join(out_dir, directory)):
+                if exists(join(out_dir, directory, 'job_completed')):
+                    # this step completed successfully.
+                    self.skip_steps.append(directory)
+                else:
+                    # work stopped before this job could be completed.
+                    rmtree(join(out_dir, directory))
+
+    def execute_pipeline(self):
+        '''
+        Executes steps of pipeline in proper sequence.
+        :return: None
+        '''
+        if not self.is_restart:
+            self.pre_check()
+
+        # this is performed even in the event of a restart.
+        self.generate_special_map()
+
+        # even if a job is being skipped, it's being skipped because it was
+        # determined that it already completed successfully. Hence,
+        # increment the status because we are still iterating through them.
+
+        self.update_status("Converting data", 1, 9)
```

Comment:
Should the 1 here be a counter and 9 be a variable so we updates in the future are easier?

Reply:
We had that at one point but it meant defining the sequence of string messages in a separate location. It became much more difficult and error-prone to update them after adding or changing a step. Now it's clearer and pretty easy to search on update_status in this file and make adjustments that way.